### ISSUE TITLE
allow parallel/multithreaded/multipart transfers between remotes

### ISSUE THREAD

**Issue by jorjao81** [rclone/rclone#7056]
#### What is your current rclone version (output from `rclone version`)?
v1.63.0

#### What problem are you are trying to solve?
Specifically, transfers from S3 to S3 are much slower than they have to be. In our case, the buckets need different credentials, so remote copying is not an option (however, in our experience remote copy is actually *slower* than local copy, and it is also not done in parallel). This happens sequentially, resulting for us in speeds of about 70MiB/s. With multithreading we can download at speeds limited only by disk and network speed. Same thing for upload using multipart.
So, actually, download to local and re-uploading is faster than a direct copy, but still half the speed we should have; AND requires more fumbling with infrastructure as the disk needs to be big enough for holding the files (> TiB).

I would like to make the in-memory copy limited only by network (and maybe cpu/mem). Our main problem is S3, but I think the solution should be at least somewhat generic.

See related forum post: https://forum.rclone.org/t/s3-s3-large-file-copy/18574/4

#### How do you think rclone should be changed to solve that?

I would actually be happy to receive some opinions on how to go about it ;-)

My initial though would be to change (or create a new) the multithreaded downloading to make it more similiar to how multipart uploads work. Instead of simply selecting threads and then using `filesize/threads` to get ranges AND threads,  we would us `chunk-size` for ranges and `download-concurrency` for number of threads.

Then we could plug this into the multi-part uploading by selecting the same sizes.

But I'm fuzzy on the details, and on how to make it more generic.

Is this a feasible idea? Any other approaches that might be better than the one I suggested?

**Comment by ncw** [comment]
Now that we've merged the first part of this - thank you @vitorog and @jorjao81 I've had a go at the next part.

This is making the multipart upload generic.

I'd really appreciate testing on this.

[v1.64.0-beta.7247.5bd9e6527.multithread](https://beta.rclone.org/branch/multithread/v1.64.0-beta.7247.5bd9e6527.multithread/) on branch [multithread](https://github.com/rclone/rclone/tree/multithread) (uploaded in 15-30 mins)

The main changes I've made are:

- pool the memory use - this is really important for not exploding memory use
- factor a multipart uploader to lib/multipart from backend/s3
- tidy and polish the ChunkWriter code
- don't do buffering when writing to local disk

This takes all the memory management out of the s3 backend which is what I wanted to achieve - that will also fix #7141

I'd appreciate it if you would test and/or review this branch.

I want to implement ChunkWriter in the following backends for v1.64 when we are sure that the interface is good.

- `azureblob`
- `b2`
- `box`
- `gcs`
- `oracleobjectstorage` ( @msays2000 maybe you'd like to help with this? )

Other possibilities

- `chunker` (see #5041)
- `dropbox` (currently not parallel uploading but there is an interface)

Still to do:

- I note also that we probably need to implement this in some of the Overlay file systems (eg `combine`, `union`) - the integration tests are failing at the moment because of the missing implementations.
- Accounting is still broken!

**Comment by Rootax** [comment]
Good job everyone ! I really hope someone can make this work with Dropbox (I mean the // upload part), since mono-chunck upload is kind of slow.

**Comment by jorjao81** [comment]
> Now that we've merged the first part of this - thank you @vitorog and @jorjao81 I've had a go at the next part.
> 
> This is making the multipart upload generic.
> 
> I'd really appreciate testing on this.
> 
> [v1.64.0-beta.7247.5bd9e6527.multithread](https://beta.rclone.org/branch/multithread/v1.64.0-beta.7247.5bd9e6527.multithread/) on branch [multithread](https://github.com/rclone/rclone/tree/multithread) (uploaded in 15-30 mins)
> 
> The main changes I've made are:
> 
> * pool the memory use - this is really important for not exploding memory use
> * factor a multipart uploader to lib/multipart from backend/s3
> * tidy and polish the ChunkWriter code
> * don't do buffering when writing to local disk
> 
> This takes all the memory management out of the s3 backend which is what I wanted to achieve - that will also fix #7141
> 
> I'd appreciate it if you would test and/or review this branch.
> 
> I want to implement ChunkWriter in the following backends for v1.64 when we are sure that the interface is good.
> 
> * `azureblob`
> * `b2`
> * `box`
> * `gcs`
> * `oracleobjectstorage` ( @msays2000 maybe you'd like to help with this? )
> 
> Other possibilities
> 
> * `chunker` (see [chunker: add support for multi-thread uploads #5041](https://github.com/rclone/rclone/issues/5041))
> * `dropbox` (currently not parallel uploading but there is an interface)
> 
> Still to do:
> 
> * I note also that we probably need to implement this in some of the Overlay file systems (eg `combine`, `union`) - the integration tests are failing at the moment because of the missing implementations.
> * Accounting is still broken!

Does it make sense to implement something for gcs? I looked into it and my understanding was that the GCS multipart upload implementation is actually the S3 protocol. So it should already be possible to configure a GCS bucket with the S3 backend, and it would just work. Do you know something I dont, or is your point more about user friendliness (I for one would prefer to configure my GCS bucket for the gcs provider, and let it dynamically choose the S3-compatible API if that makes sense)?

I would be up to implementing at least one of the listed backends, maybe azure or Dropbox. Regarding Dropbox, is the interface you meant https://github.com/dropbox/dropbox-sdk-go-unofficial/blob/master/v6/dropbox/files/client.go#L453 ?

**Comment by ncw** [comment]
> Does it make sense to implement something for gcs? I looked into it and my understanding was that the GCS multipart upload implementation is actually the S3 protocol. So it should already be possible to configure a GCS bucket with the S3 backend, and it would just work.

Yes you are right. Here is the API which says it is S3 compatible https://cloud.google.com/storage/docs/multipart-uploads - we don't currently use this so it would be quite a lot of work to add it.

I think I'll scratch GCS off the list for the time being, especially since the SDK I'm using only supports the JSON interface.

> I would be up to implementing at least one of the listed backends, maybe azure or Dropbox. Regarding Dropbox, is the interface you meant https://github.com/dropbox/dropbox-sdk-go-unofficial/blob/master/v6/dropbox/files/client.go#L453 ?

Here is the interface https://www.dropbox.com/developers/documentation/http/documentation#files-upload_session-start - you enable it with `UploadSessionType.concurrent`. We use that interface already...

Note that the dropbox uploader is very complicated due to the batching but if you want to have a go that would be great! Note that this is part of https://github.com/rclone/rclone/issues/5156

I did the azureblob interface as a second system and that went reasonably well...

New branch and betas for testing here with s3 and azureblob support.

[v1.64.0-beta.7248.c24af0144.multithread](https://beta.rclone.org/branch/multithread/v1.64.0-beta.7248.c24af0144.multithread/) on branch [multithread](https://github.com/rclone/rclone/tree/multithread) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I have pushed what I hope is the final version of this here with s3, azureblob and b2 support.

I had to change the `ChunkWriter` interface - it needed a `ctx` parameter on `WriteChunk` otherwise we can't cancel pending uploads properly.

Any testing much appreciated :-)

[v1.64.0-beta.7261.996754745.multithread](https://beta.rclone.org/branch/multithread/v1.64.0-beta.7261.996754745.multithread/) on branch [multithread2](https://github.com/rclone/rclone/tree/multithread2) (uploaded in 15-30 mins)

**Comment by Rootax** [comment]
I don't have any of theses providers but congrats !

**Comment by vitorog** [comment]
hi, @ncw , I executed some tests for S3 on an EC2 instance (linux) and on my local machine (mac) and had no issues. 🎉 🎉 


EC2 m5n.2xlarge instance (100 GB file):
```
./rclone copyto -vv --progress --s3-no-check-bucket  --disable copy --multi-thread-cutoff 1 --multi-thread-streams 100 --s3-chunk-size=25M --multi-thread-chunk-size=25M <bucket>/test-100-GiB <bucket>/TestVideo -I
...
2023-08-22 07:35:15 DEBUG : test-100-GiB: Starting multi-thread copy with 4096 parts of size 25Mi with 100 parallel streams
2023-08-22 07:35:15 DEBUG : test-100-GiB: multi-thread copy: stream 27/4096 (681574400-707788800) size 25Mi starting
2023-08-22 07:35:15 DEBUG : test-100-GiB: multi-thread copy: stream 28/4096 (707788800-734003200) size 25Mi starting
...
2023-08-22 07:38:43 DEBUG : test-100-GiB: Finished multi-thread copy with 4096 parts of size 25Mi
2023-08-22 07:38:43 DEBUG : Testvideo: Dst hash empty - aborting Src hash check
2023-08-22 07:38:43 DEBUG : test-100-GiB: Src hash empty - aborting Dst hash check
2023-08-22 07:38:43 INFO  : test-100-GiB: Multi-thread Copied (replaced existing) to: Testvideo
Transferred:   	      100 GiB / 100 GiB, 100%, 519.824 MiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      3m27.9s
2023/08/22 07:38:43 INFO  :
Transferred:   	      100 GiB / 100 GiB, 100%, 519.824 MiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      3m27.9s
...
./rclone --version
rclone v1.64.0-beta.7261.996754745.multithread2
- os/version: amazon 2 (64 bit)
- os/kernel: 4.14.320-243.544.amzn2.x86_64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.21.0
- go/linking: static
- go/tags: none
```

Local machine (1 GB file):
```
./rclone copyto --log-level DEBUG --progress --s3-no-check-bucket --multi-thread-cutoff=0   <bucket>/test-1-GiB <bucket>/test-1-GiB-test -I --multi-thread-streams 20 --s3-upload-cutoff=0 --disable=copy
...
2023-08-22 09:46:32 DEBUG : test-1-GiB: Starting multi-thread copy with 205 parts of size 5Mi with 20 parallel streams
2023-08-22 09:46:32 DEBUG : test-1-GiB: multi-thread copy: stream 3/205 (10485760-15728640) size 5Mi starting
2023-08-22 09:46:32 DEBUG : test-1-GiB: multi-thread copy: stream 13/205 (62914560-68157440) size 5Mi starting
...
2023-08-22 09:49:30 DEBUG : test-1-GiB: Finished multi-thread copy with 205 parts of size 5Mi
2023-08-22 09:49:30 DEBUG : test-1-GiB-test: Dst hash empty - aborting Src hash check
2023-08-22 09:49:30 DEBUG : test-1-GiB: Src hash empty - aborting Dst hash check
2023-08-22 09:49:30 INFO  : test-1-GiB: Multi-thread Copied (new) to: test-1-GiB-test
Transferred:            1 GiB / 1 GiB, 100%, 7.280 MiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      2m58.7s
2023/08/22 09:49:30 INFO  : 
Transferred:            1 GiB / 1 GiB, 100%, 7.280 MiB/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:      2m58.7s
...
./rclone version
rclone v1.64.0-beta.7261.996754745.multithread2
- os/version: darwin 13.4.1 (64 bit)
- os/kernel: 22.5.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.21.0
- go/linking: dynamic
- go/tags: cmount
```

**Comment by ncw** [comment]
I've merged all the outstanding multithread code to master now :-)

I had a [question on the forum](https://forum.rclone.org/t/improve-transfer-stats-calculation-for-multipart-uploads/41172/3) about this

> How is --multi-thread-streams different from --s3-upload-concurrency?

That is a good question and one I'm going to have to explain in the docs. Maybe the options need simplifying here.

Here is what has happened

### Old rclone

- if size below `--s3-upload-cutoff`  (default 200M) do single part upload
- otherwise do multipart upload using `--s3-concurrency` to set the concurrency used by the uploader.

### New rclone

- if size below `--s3-upload-cutoff` (default 200M) do single part upload
- if size below `--multi-thread-cutoff` (default 256M) do multipart upload using `--s3-concurrency` to set the concurrency used by the uploader.
- otherwise do multipart upload using  `--multi-thread-streams` to set the concurrency

Now that I've written that out it looks needlessly confusing.

I guess the options for rationalising it are

1. use `--s3-concurrency` instead of `--multi-thread-streams` to set the concurrency. This means each backend can set different concurrency levels.
- this is backwards compatible with people's config
- it means `--multi-thread-streams` will no longer be used (or maybe the default for backends that don't care)
- this will require changing the `ChunkWriter` interface to return the desired concurrency along with the desired chunk size
2. use `--multi-thread-streams` to set the concurrency everywhere. Neat, but doesn't allow different concurrency settings for different backends.
- this is not backwards compatible with people's config
- this will require reading the value of `--multi-thread-streams` in the backend (easy)
3. use a hybrid - use the maximum of `--s3-concurrency` and `--multi-thread-streams` - this allows individual control for each backend but has a global override
- this is backwards compatible with people's config
- it means `--multi-thread-streams` still does something useful
- this will require changing the `ChunkWriter` interface

Any thoughts?

PS I've pushed the 1.64 milestone back 2 weeks to allow a bit more time for this feature to settle!

**Comment by msays2000** [comment]
if download and upload speed differ we can adjust the parallelisms in source and target to match their speeds up. 
example if source data in usa and destination is england, we can have more threads to read and less concurrency to write.

**Comment by jahands** [comment]
Is there any benefit to that? Like if the concurrency is the same for reads and writes, would we just have some idle write workers while it’s waiting on reads?

Limiting write concurrency seems like it would be done implicitly because it has workers waiting on the slow read backend. But I might be missing something

**Comment by msays2000** [comment]
you are right, we can have same number of threads in each for simplicity and it will work.

**Comment by ncw** [comment]
I've merged that to master now !

**Comment by jahands** [comment]
One thing I noticed is that a larger `--multi-thread-chunk-size` does not override the backend-specific setting. Is this intentional?

**Comment by jahands** [comment]
Okay never mind, looks like `--multi-thread-chunk-size` just isn't used for s3

**Comment by ncw** [comment]
> Okay never mind, looks like `--multi-thread-chunk-size` just isn't used for s3 context: Using `provider = Scaleway`

The backend overrides it - it uses its own setting `--s3-chunk-size`.

**Comment by jahands** [comment]
Should `--multi-thread-chunk-size` be used if the backend does not specify one?

**Comment by devnoname120** [comment]
@ncw I haven't read the details of the implementation yet but can you confirm that https://github.com/rclone/rclone/pull/6179 should be rewritten to use the new `ChunkWriter` abstraction?

**Comment by devnoname120** [comment]
Also, is the current issue (#7056) a duplicate of [#6179](https://github.com/rclone/rclone/pull/6179)?

**Comment by ncw** [comment]
@devoname
> I haven't read the details of the implementation yet but can you confirm that #6179 should be rewritten to use the new `ChunkWriter` abstraction?

Only if the chunks can be uploaded in parallel and in any order. I think this is the case! The performance gains only start to show when rclone isn't filling the upload network pipe and that requires a very fast source and a big upload pipe. For your average SSD to average upload I don't think you'll notice much speed difference.


> Also, is the current issue (#7056) a duplicate of #4798?

Yes it looks like it! I'll close that issue now.

**Comment by devnoname120** [comment]
@ncw It should help if the remote has a per-connection speed throttle right? (Can be by design or by technical limitation.)

This is why I implemented a parallel chunks upload for webdav in #6179. The upload speed per connection was too low. If I recall correctly in this PR adding parallel upload workers ×4'ed the upload speed in my case (sent from the local filesystem).

**Comment by ncw** [comment]
@devoname wrote:
> It should help if the remote has a per-connection speed throttle right? (Can be by design or by technical limitation.)

The parallel uploads we have already will work already if the remote with the throttle is the destination. If the remote with the throttle is the source then this will help.

> This is why I implemented a parallel chunks upload for webdav in #6179. The upload speed per connection was too low. If I recall correctly in this PR adding parallel upload workers ×4'ed the upload speed in my case (sent from the local filesystem).

:+1:

**Comment by ncw** [comment]
This issue is complete now with many backends supporting multi thread transfers
